### ISSUE TITLE
type must be Nil, not (Int32 | Int64 | Nil)

### ISSUE THREAD

**Issue by jwoertink** [Crecto/crecto#145]
I don't quite understand the error, so I'm not sure if I'm doing something wrong here...
Here's the error
```
in lib/crecto/src/crecto/repo.cr:510: instantiating 'Array(Symbol)#each()'

      preloads.each do |preload|
               ^~~~

in lib/crecto/src/crecto/repo.cr:513: instantiating 'has_many_preload((Array(Photo) | Array(Team) | Array(TeamEvent) | Array(TeamRole) | Array(User)), Crecto::Model+:Class, Symbol)'

          has_many_preload(results, queryable, preload)
          ^~~~~~~~~~~~~~~~

in lib/crecto/src/crecto/repo.cr:532: instantiating 'join_through((Array(Photo) | Array(Team) | Array(TeamEvent) | Array(TeamRole) | Array(User)), Crecto::Model+:Class, Symbol)'

        join_through(results, queryable, preload)
        ^~~~~~~~~~~~

in lib/crecto/src/crecto/repo.cr:551: type must be Nil, not (Int32 | Int64 | Nil)

      ids = results.map(&.pkey_value.as(PkeyValue))
                          ^
```

I'm using Crecto with Kemal, so here's the bit that blows up

```crystal
class Team < Crecto::Model
  schema "teams" do
    field :id, String, primary_key: true
    field :name, String
    field :invite_code, String
    field :color, String
    field :notifications, Json
    belongs_to :user, User
    has_many :events, TeamEvent
  end

  def self.all
    Repo.all(Team)
  end
end

  get "/teams" do |env|
    # If I comment out these two lines, it works fine
    teams = Team.all
    teams.to_json    
  end
```

I'm using Crecto 0.8.4 with Crystal 0.24.1

**Comment by fridgerator** [comment]
Is `TeamEven` a join table?  im not sure why its trying to preload through a join

**Comment by jwoertink** [comment]
Nope. You just can't name a class `Event` in crystal, so I had to call it `TeamEvent`. It's just a normal one to many association. Nothing weird other than the name.

**Comment by jwoertink** [comment]
Oh, here's the `TeamEvent` model

```crystal
class TeamEvent < Crecto::Model
  schema "events" do
    field :id, String, primary_key: true
    field :name, String
    field :type, Int16
    belongs_to :team, Team    
  end
end
```

**Comment by jwoertink** [comment]
Ok, I think I may know what this is. All of my IDs are strings, which means the foreign keys are also string. I just [saw this](https://github.com/Crecto/crecto/blob/79f8a71a53eccb84ec79eb45e4d95b8f12018bac/src/types.cr#L8) which sets the foreign_key [here](https://github.com/Crecto/crecto/blob/b2f50015aa0f5b37fa06c3b39674c4336cd84257/src/crecto/schema/belongs_to.cr#L31). Anyway, that's my guess. The error may be a bit misleading, but hopefully that leads to a clue.

**Comment by fridgerator** [comment]
if you change `PkeyValue` definition in `lib/crecto/src/crecto/types.cr` to `alias PkeyValue = String | Int32 | Int64 | Nil`, does this work?

I made some more tests with strings as primary keys, and I definitely got it to fail however my error wasn't the same as yours.

**Comment by jwoertink** [comment]
ok, so looking back over this, I had to update to the master branch of crystal to fix a kemal error, but now I have another kemal and pg error. I fixed those locally, but now I'm getting some other weird things. `instance variable '@user_id' of Team must be (String | Nil), not (Int32 | Int64 | String | Nil)`

I think I want to hold off until the next release of crystal and all these shards, then give this another shot. Chances are this might not be fully crecto related anyway lol. Glad you at least found something to work out though!
